python/plugins/processing/algs/qgis/Heatmap.py

```diff
@@ -79,74 +71,116 @@ def tags(self):
+                                                         QgsProcessingParameterField.Numeric,
+                                                         optional=True
+                                                         )
         weight_field_param.setFlags(weight_field_param.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
         self.addParameter(weight_field_param)
-        kernel_shape_param = ParameterSelection(self.KERNEL,
-                                                self.tr('Kernel shape'), self.KERNELS)
+
+        keys = list(self.KERNELS.keys())
+        kernel_shape_param = QgsProcessingParameterEnum(self.KERNEL,
+                                                        self.tr('Kernel shape'),
+                                                        keys,
+                                                        allowMultiple=False,
+                                                        defaultValue=0)
         kernel_shape_param.setFlags(kernel_shape_param.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
         self.addParameter(kernel_shape_param)
-        decay_ratio = ParameterNumber(self.DECAY,
-                                      self.tr('Decay ratio (Triangular kernels only)'),
-                                      -100.0, 100.0, 0.0)
+
+        decay_ratio = QgsProcessingParameterNumber(self.DECAY,
+                                                   self.tr('Decay ratio (Triangular kernels only)'),
+                                                   QgsProcessingParameterNumber.Double,
+                                                   0.0, True, -100.0, 100.0)
         decay_ratio.setFlags(decay_ratio.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
         self.addParameter(decay_ratio)
-        output_scaling = ParameterSelection(self.OUTPUT_VALUE,
-                                            self.tr('Output value scaling'), self.OUTPUT_VALUES)
+
+        keys = list(self.OUTPUT_VALUES.keys())
+        output_scaling = QgsProcessingParameterEnum(self.OUTPUT_VALUE,
+                                                    self.tr('Output value scaling'),
+                                                    keys,
+                                                    allowMultiple=False,
+                                                    defaultValue=0)
         output_scaling.setFlags(output_scaling.flags() | QgsProcessingParameterDefinition.FlagAdvanced)
         self.addParameter(output_scaling)
-        self.addOutput(OutputRaster(self.OUTPUT_LAYER,
-                                    self.tr('Heatmap')))
 
-    def name(self):
-        return 'heatmapkerneldensityestimation'
-
-    def displayName(self):
-        return self.tr('Heatmap (Kernel Density Estimation)')
+        self.addParameter(QgsProcessingParameterRasterDestination(self.OUTPUT_LAYER, self.tr('Heatmap')))
+        self.addOutput(QgsProcessingOutputRasterLayer(self.OUTPUT_LAYER, self.tr('Heatmap')))
```

Comment:
This is no longer required - since 058271b0c1a1c7cee94c6aae7662c682cd09a410 the output will automatically be created